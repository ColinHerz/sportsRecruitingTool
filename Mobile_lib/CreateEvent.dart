import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'models/EventInfo.dart';
import 'Session.dart';
import 'dart:convert';
import 'dart:async';

class CreateEvent extends StatefulWidget{
  _CreateEvent createState() => _CreateEvent();
}

class _CreateEvent extends State<CreateEvent>{
  final _eventNameController = TextEditingController();
  final _courseController = TextEditingController();
  final _holeController = TextEditingController();
  List<TextEditingController> _parsController;
  List<Widget> _holeList;
  List<Widget> _playerList;
  List<TextEditingController> _controllerList;
  String _numPlayers = '2';
  DateTime selectedStartDate = DateTime.now();
  DateTime selectedEndDate = DateTime.now();
  List<String> listPlayers = new List<String>();

  Future<EventInfo> event(dynamic info) async {
    final http.Response _response = await http.post(
      'http://localhost3000.us-east-2.elasticbeanstalk.com/api/golf/createGolfEvent',
      headers: headers,
      body: info
    );

    if (_response.statusCode == 200) {
      Navigator.pop(context);
      //return EventInfo.fromJson(json.decode(_response.body));
    }
    else {
      print(_response.statusCode);
      print(_response.body);
    }
  }

/*
  void createHoles(int holeCount){
    _parsController = new List<TextEditingController>(holeCount);
    _holeList = new List<Widget>();

    for (int i = 0; i < holeCount; i++)
    {
      setState(() {
        _holeList.add(Expanded(
          child: Container(
            width: 70.0,
            child: TextField(
                style: TextStyle(
                    fontSize: 18
                ),
                textAlign: TextAlign.center,
                decoration: InputDecoration(
                  border: OutlineInputBorder(),
                  labelText: (i+1).toString(),
                )
            ),
          )
        )
        );
      });
    }
  }

  Text areParsDisplayed(){
    if(_holeController.text != '')
    {
      if (int.parse(_holeController.text) > 0)
        return new Text("Pars:", textAlign: TextAlign.center, style: TextStyle(fontSize: 20));
      else
        return new Text("");
    }
    else
      return new Text("");
  }

  Row parsForEachHole(){
    if (_holeController.text == '')
      return new Row();

    if (_holeList != null)
    {
        return new Row(children: _holeList);
    }
    else
      return new Row();
  }*/

  Row displayPlayers(int playerCount){
    _controllerList = new List<TextEditingController>();
    _playerList = new List<Widget>();

    for (int i = 0; i < playerCount; i++){
      _controllerList.add(new TextEditingController());

      _playerList.add(
          Expanded(child: TextField(
            controller: _controllerList[i],
              textAlign: TextAlign.center,
              decoration: InputDecoration(hintText: "Email")
          )
        )
      );
    }
    return Row(children: _playerList);
  }

  Future<void> _selectStartDate(BuildContext context) async {
    final DateTime picked = await showDatePicker(
        context: context,
        initialDate: selectedStartDate,
        firstDate: DateTime(2015, 8),
        lastDate: DateTime(2101));
    if (picked != null && picked != selectedStartDate)
      setState(() {
        selectedStartDate = picked;
      });
  }
  Future<void> _selectEndDate(BuildContext context) async {
    final DateTime picked = await showDatePicker(
        context: context,
        initialDate: selectedEndDate,
        firstDate: DateTime(2015, 8),
        lastDate: DateTime(2101));
    if (picked != null && picked != selectedEndDate)
      setState(() {
        selectedEndDate = picked;
      });
  }

  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Create Event'),
      ),
      body: Center(
        child: ListView(
          shrinkWrap: true,
          children: <Widget>[
            Padding(
                padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                child: TextField(
                    controller: _eventNameController,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Event Name',
                    )
                )
            ),
            Padding(
                padding: EdgeInsets.fromLTRB(0, 0, 0, 10),
                child: TextField(
                    controller: _courseController,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Course',
                    )
                )
            ),
            Text("${selectedStartDate.toLocal()}".split(' ')[0],
              textAlign: TextAlign.center,
              style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
            ),
            Align(
              child: RaisedButton(
                onPressed: (){
                  _selectStartDate(context);
                },
                child: Text("Select Start Date"),
              ),
            ),
            Text("${selectedEndDate.toLocal()}".split(' ')[0],
              textAlign: TextAlign.center,
              style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
            ),
            Align(
              child: RaisedButton(
                onPressed: (){
                  _selectEndDate(context);
                },
                child: Text("Select End Date"),
              ),
            ),
            /*Container(
              width: 70.0,
              child: TextField(
                style: TextStyle(
                  fontSize: 17
                ),
                textAlign: TextAlign.center,
                  controller: _holeController,
                  onSubmitted: (_holeController) {
                    setState(() {
                      createHoles(int.parse(_holeController));
                    });
                  },
                  decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    labelText: 'Holes',
                  )
              ),
            ),*/
            Padding(
              padding: EdgeInsets.fromLTRB(0, 5, 0, 5),
            ),
            //areParsDisplayed(),
            Padding(
              padding: EdgeInsets.fromLTRB(0, 5, 0, 5),
            ),
            //parsForEachHole(),
            Padding(
                padding: EdgeInsets.fromLTRB(0, 20, 0, 0),
              child: Text("Players:", textAlign: TextAlign.center ,style: TextStyle(fontSize: 20)),
            ),
            displayPlayers(int.parse(_numPlayers)),
            Center(
              child: DropdownButton<String>(
                value: _numPlayers,
                icon: Icon(Icons.arrow_downward),
                iconSize: 24,
                elevation: 16,
                style: TextStyle(color:Colors.deepOrange),
                onChanged: (String newValue) {
                  setState(() {
                    _numPlayers = newValue;
                  });
                },
                items: <String> ['2', '3', '4', '5'].map<DropdownMenuItem<String>>((String value) {
                  return DropdownMenuItem<String>(
                    value: value,
                    child: Text(value, style: TextStyle(fontSize: 18)),
                  );
                }).toList(),
              ),
            ),
            Align(
              child: RaisedButton(
                  onPressed: (){
                    for (int i = 0; i < _controllerList.length; i++)
                      listPlayers.add(_controllerList[i].text);
                    event(jsonEncode({
                      'eventName': _eventNameController.text,
                      'startDate': selectedStartDate.toIso8601String() + 'Z',
                      'endDate': selectedEndDate.toIso8601String() + 'Z',
                      'course' : _courseController.text,
                      'players' : listPlayers
                    }));
                  },
                  child: Text('Create Event')
              ),
            )
          ],
        ),
      ),
    );
  }
}
