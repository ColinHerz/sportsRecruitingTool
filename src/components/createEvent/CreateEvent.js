import apiCall from "../../api/api.js";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import React, { useState } from "react";
import { useForm } from "react-hook-form";

import "./createEvent.scss";

const NOT_ENOUGH_PLAYERS = `There must be at least one player.`;
const NOT_VALID_USER = `All users must be a valid email address.`;

const date = new Date();
const year = date.getFullYear();
const month = String(date.getMonth + 1).padStart(2, `0`);
const day = String(date.getDate() + 1).padStart(2, `0`);

const CURRENT_DATE = `${year}-${month}-${day}`;

const CreateEvent = props => {
	const { register, handleSubmit, errors } = useForm();

	const [playerName, setPlayerName] = useState(``);
	const [showPlayerInput, setShowPlayerInput] = useState(false);
	const [players, setPlayers] = useState([]);
	const [showPlayersError, setShowPlayersError] = useState(false);
	const [playersErrorText, setPlayersErrorText] = useState(``);

	const [showError, setShowError] = useState(false);

	const [submitted, setSubmitted] = useState(false);
	const [eventId, setEventId] = useState(``);

	const toggleShowPlayerInput = event => {
		event.preventDefault();

		if (showPlayerInput && playerName !== ``) {
			const newPlayers = [...players];

			setShowPlayersError(false);

			if (newPlayers.includes(playerName)) {
				return;
			}

			newPlayers.push(playerName);

			setPlayers(newPlayers);
			setPlayerName(``);
		}

		setShowPlayerInput(!showPlayerInput);
	};

	const updatePlayerName = event => {
		setPlayerName(event.target.value);
	};

	const deletePlayer = event => {
		event.preventDefault();

		const index = event.target.parentElement.id.split(`-`)[1];
		const newPlayers = [...players];

		newPlayers.splice(index, 1);

		setPlayers(newPlayers);
	};

	const onSubmit = info => {
		setShowPlayerInput(false);

		if (playersNotValid()) {
			return;
		}

		const body = Object.assign({}, info);
		body.players = players;

		apiCall(
			{
				endpoint: `/golf/createGolfEvent`,
				type: `POST`,
				body: body
			},
			data => {
				if (data.status !== 200) {
					setShowError(true);
					return;
				}

				const eid = JSON.parse(data.response).event;

				setEventId(eid);
				setSubmitted(true);
			},
			() => {
				setShowError(true);
			}
		);
	};

	const playersNotValid = () => {
		if (players.length < 1) {
			setPlayersErrorText(NOT_ENOUGH_PLAYERS);
			setShowPlayersError(true);
			return true;
		}

		if (notAllEmails()) {
			setPlayersErrorText(NOT_VALID_USER);
			setShowPlayersError(true);
			return true;
		}

		return false;
	};

	const notAllEmails = () => {
		const notEmails = players.filter(player => !player.match(/^.+@.+\..+$/));

		return notEmails.length !== 0;
	};

	if (submitted) {
		return (
			<main id="create-event-success">
				<h2>Create An Event</h2>

				<p>Event created successfully.</p>

				<p>You can view it <Link to={`/events/${eventId}`}>Here</Link></p>
			</main>
		);
	}

	return (
		<main id="create-event">
			<h2>Create An Event</h2>

			{
				showError ?
					<p className="error-text">Something went wrong. Please try again later.</p>:
					null
			}

			<form onSubmit={handleSubmit(onSubmit)}>
				<div className="input-group">
					<label>
						Event Name
						<input
							type="text"
							name="eventName"
							ref={
								register({
									required: true
								})
							}
						/>
					</label>
					{errors.eventName && <p className="error-text">An event name is required.</p>}
				</div>

				<div className="input-group">
					<label>
						Course
						<input
							type="text"
							name="course"
							ref={
								register({
									required: true
								})
							}
						/>
					</label>
					{errors.course && <p className="error-text">A course name is required.</p>}
				</div>

				<div className="input-group">
					<label>
						Start Date
						<input
							type="date"
							name="startDate"
							min={CURRENT_DATE}
							ref={
								register({
									required: true,
									min: CURRENT_DATE
								})
							}
						/>
					</label>
					{errors.startDate && <p className="error-text">A start date is required.</p>}
				</div>

				<div className="input-group">
					<label>
						End Date
						<input
							type="date"
							name="endDate"
							min={CURRENT_DATE}
							ref={
								register({
									required: true,
									min: CURRENT_DATE
								})
							}
						/>
					</label>
					{errors.endDate && <p className="error-text">An end date is required.</p>}
				</div>

				<h3>Players</h3>

				<ul>
					{
						players.map((player, index) =>
							<li
								key={index}
								id={`player-${index}`}
							>
								<span>{player}</span>
								<button onClick={deletePlayer}>Delete</button>
							</li>
						)
					}
				</ul>

				<div id="add-player-section">
					{
						showPlayerInput ?
							<input
								aria-label="New Player Name"
								type="text"
								onChange={updatePlayerName}
								value={playerName}
							/>:
							null
					}

					<button
						onClick={toggleShowPlayerInput}
						id="add-player-btn"
					>
						{
							showPlayerInput ?
								`Done`:
								`Add Player`
						}
					</button>

					{
						showPlayersError ?
							<p className="error-text">{playersErrorText}</p>:
							null
					}
				</div>

				<input type="submit" value="Create" />
			</form>
		</main>
	);
};

CreateEvent.propTypes = {
	user: PropTypes.object.isRequired
};

export default CreateEvent;
