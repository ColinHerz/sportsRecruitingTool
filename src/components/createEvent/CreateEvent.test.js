/* eslint-env jest */
import React from "react";
import renderer from "react-test-renderer";

import CreateEvent from "./CreateEvent.js";

const genProps = () => {
	return {
		user: {}
	};
};

beforeEach(() => {
	jest.resetModules();
});

describe(`Rendering`, () => {
	it(`Renders beginning page correctly`, () => {
		const props = genProps();

		const component = renderer.create(<CreateEvent {... props} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
