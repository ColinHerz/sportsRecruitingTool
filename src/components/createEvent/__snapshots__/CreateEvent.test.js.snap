// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Renders beginning page correctly 1`] = `
<main
  id="create-event"
>
  <h2>
    Create An Event
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <label>
        Event Name
        <input
          name="eventName"
          type="text"
        />
      </label>
    </div>
    <div
      className="input-group"
    >
      <label>
        Course
        <input
          name="course"
          type="text"
        />
      </label>
    </div>
    <div
      className="input-group"
    >
      <label>
        Start Date
        <input
          min="2020-function getMonth() { [native code] }1-18"
          name="startDate"
          type="date"
        />
      </label>
    </div>
    <div
      className="input-group"
    >
      <label>
        End Date
        <input
          min="2020-function getMonth() { [native code] }1-18"
          name="endDate"
          type="date"
        />
      </label>
    </div>
    <h3>
      Players
    </h3>
    <ul />
    <div
      id="add-player-section"
    >
      <button
        id="add-player-btn"
        onClick={[Function]}
      >
        Add Player
      </button>
    </div>
    <input
      type="submit"
      value="Create"
    />
  </form>
</main>
`;
